compiler/modules/types/common.bal

```diff
@@ -52,19 +59,29 @@ function memoSubtypeIsEmpty(Context cx, BddMemoTable memoTable, BddIsEmptyPredic
     int initStackDepth = cx.memoStack.length();
     cx.memoStack.push(m);
     boolean isEmpty = isEmptyPredicate(cx, b);
+    boolean cyclic = isEmpty && (m.empty == "loop");
     if !isEmpty || initStackDepth == 0 {
         foreach int i in initStackDepth + 1 ..< cx.memoStack.length() {
             MemoEmpty memoEmpty = cx.memoStack[i].empty;
-            if memoEmpty == "provisional" {
+            if memoEmpty is "provisional"|"loop"|"cyclic" {
                 cx.memoStack[i].empty = isEmpty ? isEmpty : ();
             }
         }
         cx.memoStack.setLength(initStackDepth);
         m.empty = isEmpty;
```

Comment:
m.empty = isLoop && isEmpty ? "cyclic" : isEmpty;

Reply:
The only way that we have found that this can be empty is by going through a loop. This means that the shapes in the type would all be infinite. But we define types inductively, which means we only consider finite shapes.

Reply:
fixed